pkg/controller/metrics/custom_collectors.go

```diff
@@ -328,19 +334,40 @@ func (cc clusterSyncFailingCollector) Collect(ch chan<- prometheus.Metric) {
 		return
 	}
 
+	clusterDeployments := &hivev1.ClusterDeploymentList{}
+	err = cc.client.List(context.Background(), clusterDeployments)
+	if err != nil {
+		log.WithError(err).Error("error listing cluster deployments")
+		return
+	}
+
 	for _, cs := range clusterSyncList.Items {
 		// Failing cluster syncs
 		cond := controllerutils.FindCondition(cs.Status.Conditions, hiveintv1alpha1.ClusterSyncFailed)
 		if cond != nil && cond.Status == corev1.ConditionTrue {
-			seconds := time.Since(cond.LastTransitionTime.Time).Seconds()
-			// check if duration crosses the threshold
-			if cc.minDuration.Seconds() <= seconds {
-				ch <- prometheus.MustNewConstMetric(
-					cc.metricClusterSyncFailingSeconds,
-					prometheus.GaugeValue,
-					seconds,
-					cs.Namespace+"/"+cs.Name,
-				)
+			for _, cd := range clusterDeployments.Items {
+				// Expect to find the corresponding ClusterDeployment. In the rare case it got garbage collected by the time
+				// we fetched the list of ClusterSyncs, do not report the metric.
+				if cd.Namespace == cs.Namespace {
+					// Ignore uninstalled ClusterDeployments as it wouldn't have all conditions initialized.
+					if unreachableCondition := controllerutils.FindCondition(cd.Status.Conditions, hivev1.UnreachableCondition); unreachableCondition != nil {
+						fixedLabels := make(map[string]string, len(cc.dynamicLabels.fixedLabels))
+						fixedLabels["namespaced_name"] = cs.Namespace + "/" + cs.Name
+						fixedLabels["unreachable"] = string(unreachableCondition.Status)
+						labelValues := cc.dynamicLabels.buildLabelSlice(fixedLabels, &cd)
```

Comment:
So yeah, this is uncomfortable, because we're relying on two disparate bits of code agreeing on something (the exact set of dynamic label keys) without enforcement. If we have a typo in one place or another, things could go weird, possibly in a difficult-to-notice-or-debug way.

I'll spare you where my brain went in the interim, but I think we could close the gap sufficiently by adding logic to `buildLabelSlice` to blow up if the input map contains an unrecognized key. Thoughts?

Reply:
I'm not fond of panicking while reporting a metric. I have tweaked the logic a little bit, so we will report the default value for all the labels not set in the calling method, and essentially ignore the other values that could be there. And we have the unit tests to ensure code works as expected. Does that seem acceptable?